examples/machine_learning/geneticalgorithm.cpp

```diff
@@ -0,0 +1,202 @@
+        indices = indices.rows(indices.elements() - nsamples / 2, indices.elements() - 1);
+    }
+
+    return indices;
+}
+
+
+void reproduce() {
+    //Get fittest parents
+    array selection = selectFittest();
+    array parentsx = samplex(selection);
+    array parentsy = sampley(selection);
+
+    //Divide selection in two
+    array parentsx1 = parentsx.rows(0, parentsx.elements() / 2 - 1);
+    array parentsx2 = parentsx.rows(parentsx.elements() / 2, parentsx.elements() - 1);
+    array parentsy1 = parentsy.rows(0, parentsy.elements() / 2 - 1);
+    array parentsy2 = parentsy.rows(parentsy.elements() / 2, parentsy.elements() - 1);
+
+    //Get crossover points (at which bit to crossover) and construct bit masks from them
+    array crossover = randu(nsamples / 4, u32) % (log((double)n) / log(2.0));
+    array lowermask = pow2(crossover.as(f32)).as(u32) - 1;
+    array uppermask = INT_MAX - lowermask;
+
+    //Create children as the cross between two parents
+    array children_x1 = (parentsx1 & uppermask) + (parentsx2 & lowermask);
+    array children_y1 = (parentsy1 & uppermask) + (parentsy2 & lowermask);
+
+    array children_x2 = (parentsx2 & uppermask) + (parentsx1 & lowermask);
+    array children_y2 = (parentsy2 & uppermask) + (parentsy1 & lowermask);
+
+    //Join two new sets
+    samplex = join(0, children_x1, children_x2);
+    sampley = join(0, children_y1, children_y2);
+
+    //Create mutant children
+    array mutantx = samplex.copy();
+    array mutanty = sampley.copy();
+
+    //Flip a random bit to vary the gene pool
+    int bits = (int)(log((double)n) / log(2.0));
+    mutantx = mutantx ^ (pow2((randu(nsamples / 2, u32) % bits).as(f32))).as(u32);
+    mutanty = mutanty ^ (pow2((randu(nsamples / 2, u32) % bits).as(f32))).as(u32);
+
+    samplex = join(0, samplex, mutantx);
+    sampley = join(0, sampley, mutanty);
+
+    //Update the value of each sample with the new coordinates
+    update();
+}
+
+
+void init_samples() {
+    setSeed(time(NULL));
+    samplex = randu(nsamples, u32) % n;
+    sampley = randu(nsamples, u32) % n;
+    update();
+}
+
+
+void init() {
+    //initialize space
+    searchSpace = tile(array(seq(n / 2)), 1, n / 2) + tile(moddims(array(seq(n / 2)), 1, n / 2), n / 2);
+
+    searchSpace = join(0, searchSpace, flip(searchSpace, 0));
+    searchSpace = join(1, searchSpace, flip(searchSpace, 1));
+
+    //initialize display data
+    searchSpace_x_display = iota(dim4(n, 1), dim4(1, n));
+    searchSpace_y_display = iota(dim4(1, n), dim4(n, 1));
+
+    //initalize searchers
+    init_samples();
+}
+
+void geneticSearch(bool console) {
+
+    af::info();
+    init();
+
+    float truemax = max<float>(searchSpace);
+    bool found_max = false;
+
+    if (!console) {
+        af::Window win(1600, 800, "ArrayFire Genetic Algorithm Search Demo");
+        win.grid(1, 2);
+
+        do{
+            float maximum = max<float>(samplez);
+            array where_m = where(samplez == maximum);
```

Comment:
@shehzan10 While max is not used by the algorithm, the demo displays it at every time step. Therefore a host side transfer is necessary. What do you think about a command line option to display it only when the user requests it?